### PR TITLE
NMS-12731: ResourceDao performance enhancements for the Time Series Storage API - Part 2

### PR DESCRIPTION
**Pull request by patrick-schweizer**
This PR contains performance improvements for the Timeseries Integration Layer. The overall ticket is: [NMS-12731: Optimize Performance of Timeseries Integration Layer](https://issues.opennms.org/browse/NMS-12731)

Since these are large changes I split them in multiple PRs. This PR addresses the following sub tasks:

* [ NMS-12747](https://issues.opennms.org/browse/NMS-12747): Add JMX Meters to measure and export performance
* [ NMS-12748](https://issues.opennms.org/browse/NMS-12748): Remove unnecessary writes to meta_data table
* [ NMS-12749](https://issues.opennms.org/browse/NMS-12749): Remove conversation to Newts objects in read operation
* some code cleanup

### External References

* JIRA (Issue Tracker): https://issues.opennms.org/browse/NMS-12731

### CODE REVIEW

File: features/timeseries/src/main/java/org/opennms/netmgt/timeseries/integration/persistence/TimeseriesPersistOperationBuilder.java [hunk 1]
```diff
@@ -122,8 +122,10 @@ public void setAttributeMetadata(String metricIdentifier, String name) {
 
     @Override
     public void commit() {
-        writer.insert(getSamplesToInsert());
-        writer.index(getSamplesToIndex());
+        try(final Timer.Context context = commitTimer.time()) {
+            writer.insert(getSamplesToInsert());
+            writer.index(getSamplesToIndex());
+        }
```
Comment (fooker) [inline features/timeseries/src/main/java/org/opennms/netmgt/timeseries/integration/persistence/TimeseriesPersistOperationBuilder.java:128]: This metric may be not detailed enough. I see two problems here:
* This measures to calls to the implementation which we maybe want to inspect separately.
* We also measure the time needed to build the list of samples to insert/index.

File: features/timeseries/src/main/java/org/opennms/netmgt/timeseries/integration/TimeseriesFetchStrategy.java [hunk 2]
```diff
@@ -167,108 +246,15 @@ public FetchResults fetch(long start, long end, long step, int maxrows, Long int
...
-            Utils.fillMissingValues(fetchResults, sources);
-        }
-        LOG.trace("Fetch results: {}", fetchResults);
-        return fetchResults;
+        return sourcesByResource;
     }
 
-    private Callable<Collection<Row<Measurement>>> getMeasurementsForResourceCallable(final String resourceId, final List<Source> listOfSources, final Optional<Timestamp> start, final Optional<Timestamp> end, final LateAggregationParams lag) {
-        return new Callable<Collection<Row<Measurement>>>() {
+    private Callable<Map<Source, List<Sample>>> getMeasurementsForResourceCallable(final String resourceId, final List<Source> listOfSources, final Instant start, final Instant end, final LateAggregationParams lag) {
```
[the remark this answers is not in the document]
Comment (patrick-schweizer) [reply features/timeseries/src/main/java/org/opennms/netmgt/timeseries/integration/TimeseriesFetchStrategy.java:252]: :+1: fixed

File: features/timeseries/src/main/java/org/opennms/netmgt/timeseries/integration/persistence/TimeseriesPersistOperationBuilder.java (hunk 1 above)
Comment (patrick-schweizer) [reply features/timeseries/src/main/java/org/opennms/netmgt/timeseries/integration/persistence/TimeseriesPersistOperationBuilder.java:128]: > This measures to calls to the implementation which we maybe want to inspect separately.

True, we measure the overall time to collect the samples. You think we should turn that into 2 separate timers?

> We also measure the time needed to build the list of samples to insert/index.

The actual insert happens in a different thread (managed by the ringbuffer).
We measure the insert times separately:
metadata.write.db
samples.write.ts

Comment (fooker) [reply features/timeseries/src/main/java/org/opennms/netmgt/timeseries/integration/persistence/TimeseriesPersistOperationBuilder.java:128]: Ok, if this is the collect part only, a single metric is fine

**Pull request merged**